### PR TITLE
fix(ssubscribe): properly resubscribe in case of shard failover

### PR DESCRIPTION
**Pull request by nkaradzhov**
1) when RE failover happens, there is a disconnect
    2) affected Client reconnects and tries to resubscribe all existing listeners
    ISSUE #1: CROSSSLOT Error - client was doing ssubscribe ch1 ch2.. chN which, after the failover could result in CROSSSLOT ( naturally, becasuse now some slots could be owned by other shards )
    FIX: send one ssubscribe command per channel instead of one ssubscribe for all channels
    ISSUE #2: MOVED Error - some/all of the channels might be moved somewhere else
    FIX: 1: propagate the error to the Cluster. 2: Cluster rediscovers topology.
    3: Extract all existing subscriptions from all pubsub clients and resubscribe
    over the new topology.

    fixes: #2902

### CODE REVIEW

**Comment by nkaradzhov** [comment]
run sharded pub sub tests

**Comment by uglide** [comment]
| Testcase             |   Errors |   Failures |   Skipped |   Total |
|----------------------|----------|------------|-----------|---------|
| Root Suite           |        0 |          0 |         0 |       0 |
| Sharded Pub/Sub E2E  |        0 |          0 |         0 |       0 |
| Single Subscriber    |        0 |          0 |         0 |       3 |
| Multiple Subscribers |        0 |          1 |         0 |       2 |


---- [Details for maintainers](https://github.com/redis-developer/cae-client-testing/actions/runs/18379718008)

**Comment by nkaradzhov** [comment]
run sharded pub sub tests

**Comment by uglide** [comment]
| Testcase             |   Errors |   Failures |   Skipped |   Total |
|----------------------|----------|------------|-----------|---------|
| Root Suite           |        0 |          0 |         0 |       0 |
| Sharded Pub/Sub E2E  |        0 |          0 |         0 |       0 |
| Single Subscriber    |        0 |          0 |         0 |       3 |
| Multiple Subscribers |        0 |          0 |         0 |       2 |


---- [Details for maintainers](https://github.com/redis-developer/cae-client-testing/actions/runs/18401560850)

File: packages/client/lib/cluster/cluster-slots.ts [hunk 1]
```diff
@@ -374,7 +399,9 @@ export default class RedisClusterSlots<
 
   async rediscover(startWith: RedisClientType<M, F, S, RESP>): Promise<void> {
     this.#runningRediscoverPromise ??= this.#rediscover(startWith)
-      .finally(() => this.#runningRediscoverPromise = undefined);
+      .finally(() => {
```
Comment (bobymicroby) [inline packages/client/lib/cluster/cluster-slots.ts:385]: Isn't there a potential race condition here?  If multiple _MOVED errors occur rapidly, they all get the same promise, but only the first startWith client is used. Subsequent callers might have more recent information ? I am not sure if there are going to be any implications at all, but still it made me wonder.

File: packages/client/lib/client/index.ts [hunk 2]
```diff
@@ -818,6 +818,11 @@ export default class RedisClient<
         chainId = Symbol('Socket Initiator');
 
       const resubscribePromise = this.#queue.resubscribe(chainId);
+      resubscribePromise?.catch(error => {
+        if (error.message && error.message.startsWith('MOVED')) {
+          this.emit('__MOVED')
+        }
+      });
```
Comment (vaibhavkumar-sf) [inline packages/client/lib/client/index.ts:825]: **Suggestion:** Add null/undefined check before accessing `error.message` to prevent potential runtime errors if the error object doesn't have a message property or is null/undefined. [possible bug, importance: 7]
```suggestion
const resubscribePromise = this.#queue.resubscribe(chainId);
resubscribePromise?.catch(error => {
  if (error?.message?.startsWith('MOVED')) {
    this.emit('__MOVED')
  }
});
```

File: packages/client/lib/cluster/cluster-slots.ts [hunk 3]
```diff
@@ -185,6 +185,22 @@ export default class RedisClusterSlots<
...
+    for (const master of this.masters) {
+      const shardedClient = master.pubSub?.client;
+      if (!shardedClient) continue;
+      for (const channel of shardedClient.getShardedChannels()) {
+        const listeners = shardedClient.removeShardedListeners(channel);
+        if(allChannelListeners.get(channel)) {
+          console.warn(`Found existing listeners, will be overwritten...`);
+        }
+        allChannelListeners.set(channel, listeners);
+      }
```
Comment (vaibhavkumar-sf) [inline packages/client/lib/cluster/cluster-slots.ts]: **Suggestion:** Merge listeners instead of overwriting when duplicate channels are found across shards. This prevents losing listeners and ensures all subscriptions are preserved during failover. [possible bug, importance: 9]
```suggestion
const allChannelListeners = new Map<string, ChannelListeners>();

for (const master of this.masters) {
  const shardedClient = master.pubSub?.client;
  if (!shardedClient) continue;
  for (const channel of shardedClient.getShardedChannels()) {
    const listeners = shardedClient.removeShardedListeners(channel);
    const existing = allChannelListeners.get(channel);
    if(existing) {
      existing.buffers.push(...listeners.buffers);
      existing.strings.push(...listeners.strings);
    } else {
      allChannelListeners.set(channel, listeners);
    }
  }
}
```

**Comment by vaibhavkumar-sf** [comment]
## PR Reviewer Guide 🔍

<table>
<tr><td>⏱️&nbsp;<strong>Estimated effort to review</strong>: 4 🔵🔵🔵🔵⚪</td></tr>
<tr><td>🧪&nbsp;<strong>No relevant tests</strong></td></tr>
<tr><td>🔒&nbsp;<strong>No security concerns identified</strong></td></tr>
<tr><td>⚡&nbsp;<strong>No key issues found</strong><br><br>

The PR code appears to be well-structured and follows good practices. The reviewer should still validate the logic and ensure the implementation meets requirements.</td></tr>
</table>

File: packages/client/lib/cluster/cluster-slots.ts [hunk 4]
```diff
@@ -185,6 +185,22 @@ export default class RedisClusterSlots<
...
+
+    const allChannelListeners = new Map<string, ChannelListeners>();
+
+    for (const master of this.masters) {
+      const shardedClient = master.pubSub?.client;
+      if (!shardedClient) continue;
+      for (const channel of shardedClient.getShardedChannels()) {
+        const listeners = shardedClient.removeShardedListeners(channel);
+        if(allChannelListeners.get(channel)) {
+          console.warn(`Found existing listeners, will be overwritten...`);
```
Comment (bobymicroby) [inline packages/client/lib/cluster/cluster-slots.ts]: Does this warning suggests that the same channel might exist across multiple shards? If so, this shouldn't happen in sharded pub/sub as channels should be deterministically mapped to slots. 

Does this  indicate:
  a ) A logic error in the extraction process
  b) Multiple subscriptions to the same channel (which should be consolidated)
  or c) A rare edge case during topology changes ?

**Comment by nkaradzhov** [comment]
run sharded pub sub tests

**Comment by uglide** [comment]
| Testcase   | Errors   | Failures   | Skipped   | Total   |
|------------|----------|------------|-----------|---------|


---- [Details for maintainers](https://github.com/redis-developer/cae-client-testing/actions/runs/18409985753)

**Comment by nkaradzhov** [comment]
run sharded pub sub tests

**Comment by uglide** [comment]
| Testcase             |   Errors |   Failures |   Skipped |   Total |
|----------------------|----------|------------|-----------|---------|
| Root Suite           |        0 |          0 |         0 |       0 |
| Sharded Pub/Sub E2E  |        0 |          0 |         0 |       0 |
| Single Subscriber    |        0 |          0 |         0 |       3 |
| Multiple Subscribers |        0 |          0 |         0 |       2 |


---- [Details for maintainers](https://github.com/redis-developer/cae-client-testing/actions/runs/18410445070)

**Comment by nkaradzhov** [comment]
run sharded pub sub tests

**Comment by uglide** [comment]
| Testcase             |   Errors |   Failures |   Skipped |   Total |
|----------------------|----------|------------|-----------|---------|
| Root Suite           |        0 |          0 |         0 |       0 |
| Sharded Pub/Sub E2E  |        0 |          0 |         0 |       0 |
| Single Subscriber    |        0 |          0 |         0 |       3 |
| Multiple Subscribers |        0 |          0 |         0 |       2 |


---- [Details for maintainers](https://github.com/redis-developer/cae-client-testing/actions/runs/18411166307)

**Comment by nkaradzhov** [comment]
run sharded pub sub tests

**Comment by uglide** [comment]
| Testcase             |   Errors |   Failures |   Skipped |   Total |
|----------------------|----------|------------|-----------|---------|
| Root Suite           |        0 |          0 |         0 |       0 |
| Sharded Pub/Sub E2E  |        0 |          0 |         0 |       0 |
| Single Subscriber    |        0 |          0 |         0 |       3 |
| Multiple Subscribers |        0 |          0 |         0 |       2 |


---- [Details for maintainers](https://github.com/redis-developer/cae-client-testing/actions/runs/18460033632)
